src/lib/data/data_loader.ts

```diff
@@ -1,13 +1,76 @@
-import {Budget} from '.';
-import {User} from '.';
-interface DataProvider {
-  getBudget(budgetID: string): Promise<Budget | undefined>;
-  getUserBudgets(userID: string): Promise<Budget[]>;
-  getUser(userID: string): Promise<User | undefined>;
-}
+import {Budget, Item, User} from '.';
+import {IDatabase} from './database';
+import {Collections} from '../firebase/config';
+import {Filter, Sort, Database} from './database';
+import {Abel} from 'next/font/google';
 
-interface DataModifier {
-  addBudget(budget: Budget): Promise<void>;
-}
+export class DataModel {
+  database: IDatabase;
+  constructor(database: IDatabase = Database) {
+    this.database = database;
+  }
+
+  // Get a budget by ID
+  getBudget(budgetID: string): Promise<Budget> {
+    return this.database.getDocument<Budget>(Collections.Budgets, budgetID);
+  }
+
+  getUser(userID: string): Promise<User> {
+    return this.database.getDocument<User>(Collections.Users, userID);
+  }
+  // Get a sorted and filtered list of budgets
+  getBudgets(
+    filters: Filter[] = [],
+    howMany = 25,
+    sort?: Sort
+  ): Promise<Budget[]> {
+    return this.database.getDocuments<Budget>(
+      Collections.Budgets,
+      filters,
+      sort,
+      howMany
+    );
+  }
+
+  /* getBudgetItems returns 'howMany' items in budget 'budgetID's'
+   */
+  async getBudgetItems(
+    budgetID: string,
+    sort: Sort,
+    filters: Filter[] = [],
+    howMany = 25
+  ): Promise<Item[]> {
+    filters.push(new Filter('budget_id', '==', budgetID));
+    return this.database.getDocuments<Item>(
+      Collections.Items,
+      filters,
+      sort,
+      howMany
+    );
+  }
```

Comment:
looks good! I was thinking that we could name the function `getItemsForBudget` or something like that so it's clear what the function is fetching (i.e. putting `getItems` in the first half of the name makes it more clear that you're getting items to me